consumer/transaction.go

```diff
@@ -127,27 +121,47 @@ func txnRun(s *shard, txn, prev *transaction) error {
 	return err
 }
 
-func txnBlocks(s *shard, txn, prev *transaction) bool {
+func txnBlocks(s *shard, txn *transaction) bool {
 	// We always block to await a prior commit.
 	if txn.barrierCh != nil {
 		return true
 	}
-	// We otherwise don't block if we're no longer reading messages.
-	if txn.readCh == nil {
+	// We don't block if the read channel has drained and no
+	// transaction has been started (and now cannot ever start).
+	if txn.readCh == nil && txn.consumedCount == 0 {
 		return false
 	}
```

Comment:
Ah! This is a neat detail of the Go spec and a fun technique.

A `var foo chan Bar` is always a _reference_ to an allocated channel structure, or nil. You can swap out the backing channel it points to, or set it to `nil` to nil out the reference _without touching the allocated channel_, which may have other references.

Selecting into or out of a nil channel blocks forever. It doesn't crash. That means that nil-ing out `foo` is a good way to turn it off. So, here we're not messing with the underlying message channel, we're just causing it to not select in this context.

Reply:
Ah, I see now. I didn't internalize that channels were always passed by reference. It's inconsistent since the syntax is hiding the pointer type with `var foo chan Bar` rather than `var foo *(chan Bar)` (or some syntax like that).

I knew about non-blocking reads from channels with `select`, but I don't think I put together that inside a `select` the `<-nil` would block rather than crash. Outside of `select` the `<-nil` crashes.

Reply:
Yep. `map` is also by-reference FYI.